src/wire/ipv6option.rs

```diff
@@ -0,0 +1,367 @@
+use core::fmt;
+use {Error, Result};
+
+enum_with_unknown! {
+    /// IPv6 Extension Header Option Type
+    pub doc enum Type(u8) {
+        /// 1 byte of padding
+        Pad1 =  0,
+        /// Multiple bytes of padding
+        PadN =  1
+    }
+}
+
+impl fmt::Display for Type {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            &Type::Pad1        => write!(f, "Pad1"),
+            &Type::PadN        => write!(f, "PadN"),
+            &Type::Unknown(id) => write!(f, "{}", id)
+        }
+    }
+}
+
+/// A read/write wrapper around an IPv6 Extension Header Option.
+#[derive(Debug, PartialEq)]
+pub struct Ipv6Option<T: AsRef<[u8]>> {
+    buffer: T
+}
+
+// Format of Option
+//
+// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+- - - - - - - - -
+// |  Option Type  |  Opt Data Len |  Option Data
+// +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+- - - - - - - - -
+//
+//
+// See https://tools.ietf.org/html/rfc8200#section-4.2 for details.
+mod field {
+    use wire::field::*;
+
+    // 8-bit identifier of the type of option.
+    pub const TYPE:     usize = 0;
+    // 8-bit unsigned integer. Length of the DATA field of this option, in octets.
+    pub const LENGTH:   usize = 1;
+    // Variable-length field. Option-Type-specific data.
+    pub fn DATA(length: u8) -> Field {
+        2..length as usize + 2
+    }
+}
+
+impl<T: AsRef<[u8]>> Ipv6Option<T> {
+    /// Create a raw octet buffer with an IPv6 Extension Header Option packet structure.
+    pub fn new(buffer: T) -> Ipv6Option<T> {
+        Ipv6Option { buffer }
+    }
+
+    /// Shorthand for a combination of [new] and [check_len].
+    ///
+    /// [new]: #method.new
+    /// [check_len]: #method.check_len
+    pub fn new_checked(buffer: T) -> Result<Ipv6Option<T>> {
+        let packet = Self::new(buffer);
+        packet.check_len()?;
+        Ok(packet)
+    }
+
+    /// Ensure that no accessor method will panic if called.
+    /// Returns `Err(Error::Truncated)` if the buffer is too short.
+    ///
+    /// The result of this check is invalidated by calling [set_data_length].
+    ///
+    /// [set_data_length]: #method.set_data_length
+    pub fn check_len(&self) -> Result<()> {
+        let data = self.buffer.as_ref();
+        let len = data.len();
+
+        if len < field::LENGTH {
+            return Err(Error::Truncated);
+        }
+
+        if self.option_type() == Type::Pad1 {
+            return Ok(());
+        }
+
+        let df = field::DATA(len as u8);
```

Comment:
The `len` used as input to `DATA` should be the result of `data[field::LENGTH]` instead of the buffer length. Also don't make the same mistake I did with ICMPv4 and make sure to check that the buffer is bigger than that before using that field.

Reply:
i am fixing this right now

Reply:
this is fixed now. `check_len` is tricky because of the special `Pad1` case and the variable length field.

Reply:
Yup. Thanks for the update.